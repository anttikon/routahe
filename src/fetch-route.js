import fetch from 'node-fetch'
import { get, values } from 'lodash'

const querySchema = {
  itineraries: {
    duration: {},
    legs: {
      startTime: {},
      endTime: {},
      duration: {},
      distance: {},
      mode: {},
      from: { name: {} },
      to: { name: {} },
      route: { shortName: {} },
    },
  },
}

const defaultBody = { numItineraries: 3 }

const fromAndToBody = ({ from, to }) => ({
  from: { lat: from.lat, lon: from.lon },
  to: { lat: to.lat, lon: to.lon },
})

const dateTimeToBody = ({ dateTime, arriveBy }) => {
  if (!dateTime) {
    return {}
  }
  return {
    date: dateTime.format('YYYY-MM-DD'),
    time: dateTime.format('HH:mm'),
    arriveBy,
  }
}

export const formatQueryBody = body => ({
  ...defaultBody,
  ...fromAndToBody(body),
  ...dateTimeToBody(body),
})

export const jsonToSchema = json =>
  JSON.stringify(json)
    .replace(new RegExp('"', 'g'), '')
    .replace(new RegExp(':{', 'g'), '{')
    .replace(new RegExp('{}', 'g'), '')

export const jsonToParameters = json => {
  const withoutQuotes = JSON.stringify(json)
    .replace(new RegExp('{"', 'g'), '{')
    .replace(new RegExp('":', 'g'), ':')
    .replace(new RegExp(',"', 'g'), ',')
  return `(${withoutQuotes.slice(1, -1)})`
}

export const createQuery = parameters => {
  const graphQlQuery = {
    command: 'plan',
    parameters: jsonToParameters(parameters),
    projection: jsonToSchema(querySchema),
  }
  return { query: `{${values(graphQlQuery).join('')}}` }
}

export const fetchRoute = async ({ url, body }) => {
  const result = await fetch(url, {
    method: 'POST',
    body: JSON.stringify(createQuery(body)),
    headers: { 'Content-Type': 'application/json' },
  })
  return get(await result.json(), 'data.plan.itineraries')
}
