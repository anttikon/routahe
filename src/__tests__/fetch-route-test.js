import { formatQueryBody, jsonToSchema, jsonToParameters, createQuery } from '../fetch-route'
import moment from 'moment'

const dummyBody = {
  from: {
    label: 'Tehtaankatu 24, Helsinki',
    lon: 24.944215,
    lat: 60.158061,
  },
  to: {
    label: 'Koirapuisto, Vanha-Mankkaa, Espoo',
    lon: 24.762577,
    lat: 60.181903,
  },
  dateTime: moment('2019-04-23T21:50:50.079'),
  arriveBy: undefined,
}

describe('formatQueryBody', () => {
  it('should return formatted query matching to given body without arriveBy', () => {
    expect(formatQueryBody(dummyBody)).toMatchSnapshot()
  })
  it('should return formatted query matching to given body with arriveBy', () => {
    expect(formatQueryBody({ ...dummyBody, arriveBy: true })).toMatchSnapshot()
  })
})

describe('jsonToSchema', () => {
  it('should return matching graphql schema', () => {
    const querySchema = {
      itineraries: {
        duration: {},
        legs: {
          startTime: {},
          endTime: {},
          duration: {},
          distance: {},
          mode: {},
          from: { name: {} },
          to: { name: {} },
          route: { shortName: {} },
        },
      },
    }

    expect(jsonToSchema(querySchema)).toMatchSnapshot()
  })
})

describe('jsonToParameters', () => {
  it('should return matching graphql parameters with undefined arriveBy', () => {
    const jsonParams = {
      numItineraries: 3,
      from: { lat: 60.158061, lon: 24.944215 },
      to: { lat: 60.181903, lon: 24.762577 },
      date: '2019-04-23',
      time: '21:55',
      arriveBy: undefined,
    }

    expect(jsonToParameters(jsonParams)).toMatchSnapshot()
  })

  it('should return matching graphql parameters with defined arriveBy', () => {
    const jsonParams = {
      numItineraries: 3,
      from: { lat: 60.158061, lon: 24.944215 },
      to: { lat: 60.181903, lon: 24.762577 },
      date: '2019-04-23',
      time: '21:55',
      arriveBy: true,
    }

    expect(jsonToParameters(jsonParams)).toMatchSnapshot()
  })
})

describe('createQuery', () => {
  it('should return matching query', () => {
    expect(createQuery(dummyBody)).toMatchSnapshot()
  })
})
