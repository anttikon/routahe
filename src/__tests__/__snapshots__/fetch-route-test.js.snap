// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createQuery should return matching query 1`] = `
Object {
  "query": "{plan(from:{label:\\"Tehtaankatu 24, Helsinki\\",lon:24.944215,lat:60.158061},to:{label:\\"Koirapuisto, Vanha-Mankkaa, Espoo\\",lon:24.762577,lat:60.181903},dateTime:\\"2019-04-23T18:50:50.079Z\\"){itineraries{duration,legs{startTime,endTime,duration,distance,mode,from{name},to{name},route{shortName}}}}}",
}
`;

exports[`formatQueryBody should return formatted query matching to given body with arriveBy 1`] = `
Object {
  "arriveBy": true,
  "date": "2019-04-23",
  "from": Object {
    "lat": 60.158061,
    "lon": 24.944215,
  },
  "numItineraries": 3,
  "time": "21:50",
  "to": Object {
    "lat": 60.181903,
    "lon": 24.762577,
  },
}
`;

exports[`formatQueryBody should return formatted query matching to given body without arriveBy 1`] = `
Object {
  "arriveBy": undefined,
  "date": "2019-04-23",
  "from": Object {
    "lat": 60.158061,
    "lon": 24.944215,
  },
  "numItineraries": 3,
  "time": "21:50",
  "to": Object {
    "lat": 60.181903,
    "lon": 24.762577,
  },
}
`;

exports[`jsonToParameters should return matching graphql parameters with defined arriveBy 1`] = `"(numItineraries:3,from:{lat:60.158061,lon:24.944215},to:{lat:60.181903,lon:24.762577},date:\\"2019-04-23\\",time:\\"21:55\\",arriveBy:true)"`;

exports[`jsonToParameters should return matching graphql parameters with undefined arriveBy 1`] = `"(numItineraries:3,from:{lat:60.158061,lon:24.944215},to:{lat:60.181903,lon:24.762577},date:\\"2019-04-23\\",time:\\"21:55\\")"`;

exports[`jsonToSchema should return matching graphql schema 1`] = `"{itineraries{duration,legs{startTime,endTime,duration,distance,mode,from{name},to{name},route{shortName}}}}"`;
